File: src/main/java/me/miki/shindo/management/mods/impl/ParticleCustomizerMod.java
package me.miki.shindo.management.mods.impl;

import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventAttackEntity;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.mods.Mod;
import me.miki.shindo.management.mods.ModCategory;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.potion.Potion;
import net.minecraft.util.EnumParticleTypes;

import me.miki.shindo.management.settings.config.Property;
import me.miki.shindo.management.settings.config.PropertyType;
public class ParticleCustomizerMod extends Mod {

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ALWAYS_SHARPNESS)
    private boolean alwaysSharpnessSetting = false;
    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ALWAYS_CRITICALS)
    private boolean alwaysCriticalsSetting = false;
    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.SHARPNESS)
    private boolean sharpnessSetting = true;
    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.CRITICALS)
    private boolean criticalsSetting = false;

    @Property(type = PropertyType.NUMBER, translate = TranslateText.SHARPNESS_AMOUNT, min = 1, max = 10, current = 2, step = 1)
    private int sharpnessAmountSetting = 2;
    @Property(type = PropertyType.NUMBER, translate = TranslateText.CRITICALS_AMOUNT, min = 1, max = 10, current = 2, step = 1)
    private int criticalsAmountSetting = 2;

    public ParticleCustomizerMod() {
        super(TranslateText.PARTICLE_CUSTOMIZER, TranslateText.PARTICLE_CUSTOMIZER_DESCRIPTION, ModCategory.RENDER);
    }

    @EventTarget
    public void onAttackEntity(EventAttackEntity event) {

        EntityPlayer player = mc.thePlayer;

        int sMultiplier = sharpnessAmountSetting;
        int cMultiplier = criticalsAmountSetting;

        if (!(event.getEntity() instanceof EntityLivingBase)) {
            return;
        }

        boolean critical = criticalsSetting && player.fallDistance > 0.0F && !player.onGround && !player.isOnLadder() && !player.isInWater() && !player.isPotionActive(Potion.blindness) && player.ridingEntity == null;
        boolean alwaysSharpness = alwaysSharpnessSetting;
        boolean sharpness = sharpnessSetting && EnchantmentHelper.getModifierForCreature(player.getHeldItem(), ((EntityLivingBase) event.getEntity()).getCreatureAttribute()) > 0;
        boolean alwaysCriticals = alwaysCriticalsSetting;

        if (critical || alwaysCriticals) {
            for (int i = 0; i < cMultiplier - 1; i++) {
                mc.effectRenderer.emitParticleAtEntity(event.getEntity(), EnumParticleTypes.CRIT);
            }
        }

        if (alwaysSharpness || sharpness) {
            for (int i = 0; i < sMultiplier - 1; i++) {
                mc.effectRenderer.emitParticleAtEntity(event.getEntity(), EnumParticleTypes.CRIT_MAGIC);
            }
        }
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/ParticleCustomizerMod.kt
package me.miki.shindo.management.mods.impl

import me.miki.client_api.event.EventTarget
import me.miki.shindo.management.event.impl.EventAttackEntity
import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.mods.Mod
import me.miki.shindo.management.mods.ModCategory
import me.miki.shindo.management.nanovg.font.LegacyIcon
import me.miki.shindo.management.settings.config.Property
import me.miki.shindo.management.settings.config.PropertyType
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.potion.Potion
import net.minecraft.util.EnumParticleTypes

class ParticleCustomizerMod : Mod(
    TranslateText.PARTICLE_CUSTOMIZER,
    TranslateText.PARTICLE_CUSTOMIZER_DESCRIPTION,
    ModCategory.RENDER,
    LegacyIcon.MOD_PARTICLE_CUSTOMIZER
) {
    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ALWAYS_SHARPNESS)
    private val alwaysSharpnessSetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ALWAYS_CRITICALS)
    private val alwaysCriticalsSetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.SHARPNESS)
    private val sharpnessSetting = true

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.CRITICALS)
    private val criticalsSetting = false

    @Property(
        type = PropertyType.NUMBER,
        translate = TranslateText.SHARPNESS_AMOUNT,
        min = 1.0,
        max = 1.00,
        current = 2.0,
        step = 1.0
    )
    private val sharpnessAmountSetting = 2

    @Property(
        type = PropertyType.NUMBER,
        translate = TranslateText.CRITICALS_AMOUNT,
        min = 1.0,
        max = 1.00,
        current = 2.0,
        step = 1.0
    )
    private val criticalsAmountSetting = 2

    @EventTarget
    fun onAttackEntity(event: EventAttackEntity) {
        val player: EntityPlayer = mc.thePlayer

        val sMultiplier = sharpnessAmountSetting
        val cMultiplier = criticalsAmountSetting

        if (event.getEntity() !is EntityLivingBase) {
            return
        }

        val critical =
            criticalsSetting && player.fallDistance > 0.0f && !player.onGround && !player.isOnLadder && !player.isInWater && !player.isPotionActive(
                Potion.blindness
            ) && player.ridingEntity == null
        val alwaysSharpness = alwaysSharpnessSetting
        val sharpness = sharpnessSetting && EnchantmentHelper.getModifierForCreature(
            player.heldItem,
            (event.getEntity() as EntityLivingBase).creatureAttribute
        ) > 0
        val alwaysCriticals = alwaysCriticalsSetting

        if (critical || alwaysCriticals) {
            for (i in 0 until cMultiplier - 1) {
                mc.effectRenderer.emitParticleAtEntity(event.getEntity(), EnumParticleTypes.CRIT)
            }
        }

        if (alwaysSharpness || sharpness) {
            for (i in 0 until sMultiplier - 1) {
                mc.effectRenderer.emitParticleAtEntity(event.getEntity(), EnumParticleTypes.CRIT_MAGIC)
            }
        }
    }
}





